### PR TITLE
types(ApplicationCommandManager): Deprecate old `*Data` type usages and allow camel cased dapi types to be used

### PR DESCRIPTION
**Pull request by suneettipirneni**
**Please describe the changes this PR makes and why it should be merged:**

Ref: #6958 

**Status and versioning classification:**
- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating

### CODE REVIEW

**Comment by ImRodry** [comment]
Do we really need to add a dependency for types only? Most people don't use typescript and the package offers much more than the type that is being imported, I believe we could simply create the type ourselves instead

**Comment by suneettipirneni** [comment]
> Do we really need to add a dependency for types only? Most people don't use typescript and the package offers much more than the type that is being imported, I believe we could simply create the type ourselves instead

It's a dev dep, so why does it matter. And the camel case util type here is more sophisticated. It's not that simple to implement one our own that's as robust as this one.

**Comment by ImRodry** [comment]
If it’s a dev dev then these types won’t even work at all because the package wouldn’t get installed. We can simply copy the type from them and add it ourselves

File: typings/index.d.ts [hunk 1]
```diff
@@ -137,6 +138,7 @@ import {
   RawWidgetData,
   RawWidgetMemberData,
 } from './rawDataTypes';
+import type { CamelCasedPropertiesDeep } from 'type-fest';
```
Comment (almeidx) [inline typings/index.d.ts:141]:
```suggestion
```

**Review by almeidx — CHANGES_REQUESTED** [review]

**Comment by suneettipirneni** [comment]
We won't be using our own version of Camelizer. We are going to stick with the package that was originally brought up in this PR.

File: typings/index.d.ts [hunk 2]
```diff
@@ -2627,7 +2635,11 @@ export abstract class CachedManager<K, Holds, R> extends DataManager<K, Holds, R
   private _add(data: unknown, cache?: boolean, { id, extras }?: { id: K; extras: unknown[] }): Holds;
 }
 
-export type ApplicationCommandDataResolvable = ApplicationCommandData | RESTPostAPIApplicationCommandsJSONBody;
+export type ApplicationCommandDataResolvable =
+  | Camelize<RESTPostAPIApplicationCommandsJSONBody>
+  | RESTPostAPIApplicationCommandsJSONBody;
+
+export type foo = Camelize<RESTPostAPIApplicationCommandsJSONBody>['options'];
```
Comment (almeidx) [inline typings/index.d.ts:2642]: 😕

File: tslint.json [hunk 3]
```diff
@@ -24,6 +24,7 @@
     "array-type": [true, "array"],
     "one-line": false,
     "no-any-union": false,
-    "void-return": false
+    "void-return": false,
+    "unified-signatures": false
```
Comment (Jiralite) [inline tslint.json:28]: With the introduction of `unified-signatures: false`, I believe these can be safely removed now:

https://github.com/discordjs/discord.js/blob/9d7152f0fdcd879cad654fe4fdd658fac57ceb02/typings/index.d.ts#L2341-L2345

https://github.com/discordjs/discord.js/blob/9d7152f0fdcd879cad654fe4fdd658fac57ceb02/typings/index.d.ts#L3095-L3099

**Comment by vladfrangu** [comment]
Well what do ya know! This needs a rebase

File: typings/tests.ts [hunk 4]
```diff
@@ -880,13 +898,18 @@ declare const applicationNonChoiceOptionData: ApplicationCommandOptionData & {
   // @ts-expect-error
   applicationNonChoiceOptionData.choices;
 }
+declare const applicationChoiceOptionData: APIApplicationCommandOption & { type: ApplicationCommandOptionType.String };
+{
+  // Choices should be available.
+  applicationChoiceOptionData.choices;
+}
```
Comment (vladfrangu) [inline typings/tests.ts:905]: What's the point of this test? You're just checking a discord-api-type 👀

**Review by vladfrangu — APPROVED** [review]
🎉

**Comment by iCrawl** [comment]
This needs another rebase for the new type testing PR that landed.

**Review by SpaceEEC — APPROVED** [review]

**Pull request merged**